File: tree/236_lowest_common_ancestor_of_a_binary_tree.cc
#include "tree_node.h"

// 2024/01/10: 直接抄答案

TreeNode* res;

// root的整颗树是否含有p或者q.
bool Dfs(TreeNode* root, TreeNode* p, TreeNode* q) {
  if (root == nullptr) {
    return false;
  }
  bool left_son = Dfs(root->left, p, q);
  bool right_son = Dfs(root->right, p, q);
  if ((left_son && right_son) ||
      ((root->val == p->val || root->val == q->val) &&
       (left_son || right_son))) {
    res = root;
  }
  return left_son || right_son || root->val == p->val || root->val == q->val;
}

TreeNode* LowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
  Dfs(root, p, q);
  return res;
}

File: tree/236_lowest_common_ancestor_of_a_binary_tree.py
from tree.tree_node import TreeNode, create_tree_node
from typing import Optional

res = None


# 2026/05/20: 直接抄答案.
# 判断 p 或者 q 是否在整颗树里面.
def check(
    root: Optional[TreeNode], p: Optional[TreeNode], q: Optional[TreeNode]
) -> bool:
    global res

    if root is None:
        return False

    left_son = check(root.left, p, q)
    right_son = check(root.right, p, q)
    if (left_son and right_son) or (
        (root.val == p.val or root.val == q.val) and (left_son or right_son)
    ):
        res = root

    return left_son or right_son or root.val == p.val or root.val == q.val


def lowest_common_ancestor(
    root: Optional[TreeNode], p: Optional[TreeNode], q: Optional[TreeNode]
) -> Optional[TreeNode]:
    check(root, p, q)
    return res


root = [3, 5, 1, 6, 2, 0, 8, None, None, 7, 4]
p = 5
q = 1
root = create_tree_node(root)
